### PR TITLE
feat: add support for constants/literals in vectorized map

### PR DESCRIPTION
**Pull request by onelson**
Refs #4622

## Overview

Previously, functions passed to `map()` would not vectorize when the body contained items that _were not members on the incoming record_.

This diff expands eligibility to include constant values, specifically a short list of literal value types.

Currently supported types:
- int
- float
- string (no interpolation)
- time

Notable omissions:
- uint
- boolean

Uint will be trivially supported except for the fact I don't think we have a way to write them as _as a literal_. Seems like we would need to first vectorize the `uint()` builtin.

Boolean seems like it should "just work", but I see "undefined identifier" errors in libflux when I tried to enable this.

## Implementation Notes

This work took a few turns along the way, but here's where things landed.

When literal expressions are encountered during the "vectorize" stage of the semantic graph walk, they are wrapped with a "fake" call expression, `~~vecRepeat~~(v: ...)`.

Meanwhile in the call expression evaluator, we look for functions with this name and intercept to replace the call with a new "const vector" evaluator which sets up a new `VectorRepeatValue` using the `v` argument seen in the call expression.

For there, this value will "pass" as `Vector` though it does not hold an array. This unfortunately means we need to take care with how we use it :frowning_face:

Where before we had functions that accept 2 vectors per operator/type, this PR adds _another 2 functions_ for each case. The new flavors accept a vector and a scalar value, varied by left/right sides for each. By checking the types of the vectors, we select the appropriate function to use.

For the case where both left/right sides are a `VectorRepeatValue`, we delegate to the row-based (non-vectorized) implementation on the pair of inner Values to produce a new `VectorRepeatValue` holding the result. This is sometimes referred to as "constant folding" in the comments.

When `VectorRepeatValue`s are used in binary operations, the output of that operation is an array-backed Vector, but in cases where a literal is given as a field without any such op, such as `{r with x: 123}` it's possible for a `VectorRepeatValue` to "survive" through to the end of the map2 transformation. In this situation, since map2 aims to return a collection of arrays, we need to "fill" an array of the appropriate size with the same constant value. _The implementation for this "fill" can likely be specialized/improved_.

## Additional

Included in this PR is some added checks to the vectorize eligibility for binary expressions. Part of the stdlib test suite started to fail once the new code path was added because functions like `(r) => ({ r with x: r.a > 0 })` were suddenly becoming vectorized but the `>` operator is not yet implemented. Previously these did not vectorize because of the literal, but with the fix proposed here they will now be ineligible _because of the operator_. Ref: #4608

### Done checklist
- [ ] docs/SPEC.md updated **N/A**
- [x] Test cases written

### CODE REVIEW

File: compiler/compiler.go [hunk 1]
```diff
@@ -700,6 +700,15 @@ func compile(n semantic.Node, subst semantic.Substitutor) (Evaluator, error) {
 		if err != nil {
 			return nil, err
 		}
+		// Rewrite of "fake function" to vectorize constants.
+		if c, ok := n.Callee.(*semantic.IdentifierExpression); ok && c.Name.LocalName == "~~vecRepeat~~" {
```
Comment (onelson) [inline compiler/compiler.go:704]: named `~~vecRepeat~~` so as to be illegal as a symbol coming from userland (ie the flux source itself), thus making it impossible to clash. The choice to use double-enclosing tilde was just a nod to the fact the function call was being erased and replaced: ~~vecRepeat~~.

**Review by OfTheDelmer — APPROVED** [review]
Looks great

File: array/binary.gen.go.tmpl [hunk 2]
```diff
@@ -34,15 +33,45 @@ func {{$type}}{{$op.Name}}(l, r *{{$type}}, mem memory.Allocator) (*Float, error
...
 }
 
-{{else}}
+func {{$type}}{{$op.Name}}LConst(l {{index $.TypeMap $type}}, r *{{$type}}, mem memory.Allocator) (*Float, error) {
+	n := r.Len()
+	b := NewFloatBuilder(mem)
+	b.Resize(n)
+	for i := 0; i < n; i++ {
+		if r.IsValid(i) {
+			b.Append(math.Pow(float64(l), float64(r.Value(i))))
```
[the remark this answers is not in the document]
Comment (onelson) [reply array/binary.gen.go.tmpl:42]: I didn't think about it. Happy to hoist the cast out though. Seems like the correct thing to do in this situation. :+1:

File: values/binary.gen.go.tmpl [hunk 3]
```diff
@@ -9,9 +9,23 @@ import (
...
 func vector{{$op.Name}}(l, r Vector, mem memory.Allocator) (Value, error) {
+	var lvr, rvr *Value
+	if vr, ok := l.(*VectorRepeatValue); ok {
+		lvr = &vr.val
+	}
+	if vr, ok := r.(*VectorRepeatValue); ok {
+		rvr = &vr.val
+	}
+
+	if lvr != nil && rvr != nil {
```
[the remark this answers is not in the document]
Comment (onelson) [reply values/binary.gen.go.tmpl:23]: Hopefully the phrasing in that panic makes sense. Handling this a layer above was just much more convenient.

**Pull request merged**